### PR TITLE
DD4hepSim : parseOptions(argv) customize

### PR DESCRIPTION
**Pull request by rete**
BEGINRELEASENOTES
- Modified DD4hepSimulation.parseOptions() to receive argv argument, allowing user scripts to pass custom command line args 
ENDRELEASENOTES

### CODE REVIEW

**Comment by rete** [comment]
Pull-requested to early... I need modifications. Sorry ... do not merge

**Comment by andresailer** [comment]
I don't think any of this necessary, you can directly modify `sys.argv` in your script before calling `parseOptions`.

**Comment by gaede** [comment]
Andre, we had discussed this option but came to the conclusion that it is much cleaner to do it like this, as the calling program might also have it's own set of argv that than will have to be parsed before (and not accessed later in the code anywhere )....

File: DDSim/DDSim/DD4hepSimulation.py [hunk 1]
```diff
@@ -108,16 +110,18 @@ def readSteeringFile(self):
         self.__dict__ = obj.__dict__
     self.steeringFile = os.path.abspath(sFileTemp)
 
-  def parseOptions(self):
+  def parseOptions(self, argv=sys.argv):
```
Comment (andresailer) [inline DDSim/DDSim/DD4hepSimulation.py]: This is bad due to the way python treats default arguments.
Use `None` as the default argument and then
```
if argv is None:
  self._argv = list(sys.argv)
```

File: DDSim/DDSim/DD4hepSimulation.py [hunk 2]
```diff
@@ -108,16 +110,18 @@ def readSteeringFile(self):
...
     """parse the command line options"""
+    self.argv = argv
```
Comment (andresailer) [inline DDSim/DDSim/DD4hepSimulation.py]: This just assigns the "reference" to argv so `self.argv` and `argv` are referring to the same object.
Should do
`self._argv = list(argv)` to avoid surprises later
```
>>> l = []
>>> b = l
>>> b.append("arg")
>>> l
['arg']
>>> c = list(l)
>>> c
['arg']
>>> l.append("arg2")
>>> c
['arg']
>>> l
['arg', 'arg2']
```

File: DDSim/DDSim/DD4hepSimulation.py [hunk 3]
```diff
@@ -635,7 +639,7 @@ def __setupSensitiveDetectors(self, detectors, setupFuction, defaultFilter=None)
       for pattern in self.action.mapActions:
         if pattern.lower() in det.lower():
           action = self.action.mapActions[pattern]
-          print  '       replace default action with : ' , action 
+          print  '       replace default action with : ' , action
```
Comment (andresailer) [inline DDSim/DDSim/DD4hepSimulation.py]: Unnecessary whitespace change, please exclude from your commit

File: DDSim/DDSim/DD4hepSimulation.py [hunk 4]
```diff
@@ -89,6 +89,8 @@ def __init__(self):
     self.filter = Filter()
     self.physics = Physics()
 
+    self.argv = None
```
Comment (andresailer) [inline DDSim/DDSim/DD4hepSimulation.py]: use 
```
self._argv
```
because, for example, we don't need this to show up the steering file when dumped

**Comment by andresailer** [comment]
Please don't forget to squash your commits and remove the "WIP" from the title when you are done.

**Comment by andresailer** [comment]
Please squash your commits

**Comment by rete** [comment]
I followed the guideline from the CONTRIBUTING.md file (NoSY) but I didn't know what squashing commits mean so I didn't do it. I merged my local branch into the master, pushed it to my fork and deleted my local branch.
Now I don't know how to squash them. Sorry...

**Comment by andresailer** [comment]
Ahh, you brought your branch back.
```
git checkout ddsim-argv-parsing
git branch backup-parsing # just in case make a "copy" of the current branch
git reset --hard 0a5d736
git push -f downstream ddsim-argv-parsing ## depends what you call your remote
```
